src/mon/PGMonitor.cc

```diff
@@ -2223,3 +2231,78 @@ void PGMonitor::check_sub(Subscription *sub)
 		    sub->session->con.get());
   }
 }
+
+void PGMonitor::get_pg_state(stringstream& ds, vector<string>& states, ceph::unordered_map<pg_t, pg_stat_t>& pg_stats, 
+                             const int64_t poolid, const int64_t osdid)
+{
+  __u32 type = 0;
+  if (states[0] == "active")
+    type = PG_STATE_ACTIVE;
+  else if (states[0] == "clean")
+    type = PG_STATE_CLEAN;
+  else if (states[0] == "down")
+    type = PG_STATE_DOWN;
+  else if (states[0] == "replay")
+    type = PG_STATE_REPLAY;
+  else if (states[0] == "splitting")
+    type = PG_STATE_SPLITTING;
+  else if (states[0] == "scrubbing")
+    type = PG_STATE_SCRUBBING;
+  else if (states[0] == "degraded")
+    type = PG_STATE_DEGRADED;
+  else if (states[0] == "inconsistent")
+    type = PG_STATE_INCONSISTENT;
+  else if (states[0] == "peering")
+    type = PG_STATE_PEERING;
+  else if (states[0] == "recoverying")
+    type = PG_STATE_RECOVERING;
+  else if (states[0] == "backfill_wait")
+    type = PG_STATE_BACKFILL_WAIT;
+  else if (states[0] == "incomplete")
+    type = PG_STATE_INCOMPLETE;
+  else if (states[0] == "remapped")
+    type = PG_STATE_INCOMPLETE;
+  else if (states[0] == "stale")
+    type = PG_STATE_STALE;
+  else if (states[0] == "deep_scrub")
+    type = PG_STATE_DEEP_SCRUB;
+  else if (states[0] == "backfill")
+    type = PG_STATE_DEEP_SCRUB;
+  else if (states[0] == "backfill_toofull")
+    type = PG_STATE_BACKFILL_TOOFULL;
+  else if (states[0] == "recovery_wait")
+    type = PG_STATE_RECOVERY_WAIT;
+  else if (states[0] == "undersized")
+    type = PG_STATE_UNDERSIZED;
+  else {
+    if (states[0] != "all")
+      ds << "Unknown type: " << type << std::endl;
+  }
+  for (ceph::unordered_map<pg_t, pg_stat_t>::const_iterator i = pg_map.pg_stat.begin();
+       i != pg_map.pg_stat.end();
+       ++i) {
+     if ( poolid >= 0) {
+       if (i->first.pool() != uint64_t(poolid))
+         continue;
+       if (osdid >= 0) {
+         if (i->second.acting_primary != osdid)
+           continue;
+         if ((states[0] == "all") || (i->second.state & type))
+           pg_stats[i->first] = i->second;
+       } else {
+         if ((states[0] == "all") || (i->second.state & type))
+           pg_stats[i->first] = i->second;
+       }
+     } else {
+       if (osdid >= 0) {
+         if (i->second.acting_primary != osdid)
+           continue;
+         if ((states[0] == "all") || (i->second.state & type))
+           pg_stats[i->first] = i->second;
+       } else {
+         if((states[0] == "all") || (i->second.state & type))
+           pg_stats[i->first] = i->second;
+       }
+     }
+  }
```

Comment:
Don't you think it would be shorter with something like:
<pre>
 if ( poolid >= 0 && i->first.pool() != uint64_t(poolid))
    continue;
if (osdid >= 0 && i->second.acting_primary != osdid)
   continue
if ((states[0] == "all") || (i->second.state & type))
   pg_stats[i->first] = i->second;
</pre>